elastic/elastic-agent
# Feature/add agent version to installation directory name

## What does this PR do?
This PR uses the package manifest included in elastic-agent packages to map the package contents onto disk when installing/upgrading agent.
In the context of https://github.com/elastic/elastic-agent/issues/2579 the path mapping is used to install or upgrade agent in a directory like data/elastic-agent-<package version>-<hash> while we maintain backward compatibility for the package structure using the previous `elastic-agent-<hash>`.

This PR is the aggregation of #4173 #4174 #4176 #4177 #4178 that have been used to split the change in reviewable fragments.

With this change elastic-agent will use the path mappings present in the package manifest to remap package files before writing them on disk in the install directory.

There is no integration test in this change as that requires some hacking of the agent packages due to current limitations of integration test framework. This will be tackled in #2780 



## Why is it important?

This changes implements #2579 allowing different versions of agent built from the same git commit (for example a repackaged agent version containing patched components) to upgrade between themselves.


## Checklist




## Author's Checklist



## How to test this PR locally



To test this PR locally, we package elastic-agent twice from the same commit specifying a different AGENT_PACKAGE_VERSION like:
```
EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```
and
```
AGENT_PACKAGE_VERSION=8.13.0-repackaged EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```

Install the first agent package and verify that it's healthy and it runs from `data/elastic-agent-8.13.0-SNAPSHOT-<hash>` directory.
Upgrade this agent using the second package
```
elastic-agent upgrade 8.13.0-repackaged-SNAPSHOT --source-uri file://<directory where the package is located on disk> --skip-verify
```

Upgrade will unpack and store the new agent in `data/elastic-agent-8.13.0-repackaged-<hash>` and rotate the links.
Verify the agent directory structure contains the 2 distinct directories.

If we want to test the rollback functionality it's enough to kill the new agent a few times after the upgrade or break the new agent install in some way.

## Related issues


- Closes #2579
- Supersedes #3950 

## Use cases



## Screenshots



## Logs



## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
